### DISCUSSION TITLE
Type error when using max(iter, key)

### DISCUSSION THREAD

**Discussion by GhostlyCipher** [microsoft/pyright#7723] [Q&A]
```
from collections import Counter

buckets = ["apple", "banana", "apple", "orange", "banana", "apple"]
bucket_counts = Counter(buckets)
print(max(bucket_counts, key=bucket_counts.get))
```

```
error: No overloads for "max" match the provided arguments (reportCallIssue)
error: Argument of type "Overload[(key: str, /) -> (int | None), (key: str, default: int, /) -> int, (key: str, default: _T@get, /) -> (int | _T@get)]" cannot be assigned to parameter "key" of type "(_T@max) -> SupportsRichComparison" in function "max"
    No overloaded function matches type "(str) -> SupportsRichComparison" (reportArgumentType)
```

I think this is a perfectly valid use case and I'm wondering why this is being reported as an error.
This is the max function which I think the usage should be mapped to.

```
def max(
    iterable: Iterable[_T@max],
    /,
    *,
    key: (_T@max) -> SupportsRichComparison
) -> _T@max: ...
```

Please let me understand how to fix this

**Comment by erictraut** [comment]
Pyright's behavior here is correct given the way `max` is defined in the typeshed stubs. Mypy generates the same error in this case.

The issue is that the `Counter.get` method (the version that accepts a single argument) returns a value of `str | None`, but `None` is not comparable to `str`, so it's not compatible with `SupportsRichComparison`.

**Reply by GhostlyCipher** [reply]
Thank you for the clarification. I updated my code to this:
```
print(max(bucket_counts, key=lambda x: bucket_counts[x]))
```
This works as expected without pyright errors.
